### PR TITLE
R91

### PR DESCRIPTION
**Pull request by BenFradet**
- [x] bump eer version
- [x] test with nested dir and scalding
- [x] --resume-from
- [x] lock
- [x] disable gen commands
- [x] prep for rel
- [x] rel date in the changelog
- [ ] merge wiki
- [x] ~~upload rmr script to the standard bucket~~

### CODE REVIEW

**Comment by alexanderdean** [comment]
I'd make the Ruby bump an explicit ticket, just for clarity...

**Comment by BenFradet** [comment]
It is already: https://github.com/snowplow/snowplow/pull/3029/commits/70c9bc1dba4a9be01dcb086a9881a768deb527eb

It'll disappear with the rebase.

**Comment by alexanderdean** [comment]
Oh sorry!

**Comment by BenFradet** [comment]
- [x] run
- [ ] generate
  - [x] generate emr-cluster
  - [x] generate playbook
  - [ ] generate all
- [ ] lint
  - [ ] lint resolver
  - [ ] lint enrichments
  - [ ] lint all
- [ ] s3 tasks -> s3distcp with manifests

File: 3-enrich/emr-etl-runner/lib/snowplow-emr-etl-runner/linter.rb [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        self
+      end
+
+      # Lints an Iglu config by trying to parse the resolver config
+      Contract String => Or[Linter, LinterError]
+      def self.lint(resolver_config)
+        begin
+          puts resolver_config
+          json = JSON.parse(resolver_config, {:symbolized_names => true})
+          resolver = Iglu::Resolver.parse(json)
```
Comment (BenFradet) [inline 3-enrich/emr-etl-runner/lib/snowplow-emr-etl-runner/linter.rb]: @chuwy This lines gives me:

```
NoMethodError:
       undefined method `map' for nil:NilClass
       Did you mean?  tap
     # /home/vagrant/.rvm/gems/jruby-9.1.7.0/gems/iglu-ruby-client-0.1.0/lib/iglu-client/resolver.rb:71:in `parse'
     # ./spec/snowplow-emr-etl-runner/linter_spec.rb:35:in `block in subject'
     # ./spec/snowplow-emr-etl-runner/linter_spec.rb:40:in `block in (root)'
```

**Comment by BenFradet** [comment]
I think I'm done, reviews very welcome.

**Comment by alexanderdean** [comment]
Excellent! Requesting initial review from Anton...

File: 3-enrich/emr-etl-runner/lib/snowplow-emr-etl-runner/cli.rb [hunk 2]
```diff
@@ -153,38 +203,103 @@ def self.load_config(config_file, optparse)
...
+      # Validate our arguments against the configuration Hash
+      # Make updates to the configuration Hash based on the
+      # arguments
+      Contract ArgsHash, ConfigHash => ConfigHash
+      def self.validate_and_coalesce(args, config)
+
+        # Check our skip argument
+        unless args[:resume_from].nil?
+          unless RESUMABLES.include?(args[:resume_from])
+            raise ConfigError, "Invalid option: resume-from can be 'enrich', 'shred', 'elasticsearch', or 'archive_raw', not '#{args[:resume_from]}'"
```
Comment (chuwy) [inline 3-enrich/emr-etl-runner/lib/snowplow-emr-etl-runner/cli.rb]: Valid options can be dymamicaly generated.

File: 3-enrich/emr-etl-runner/Gemfile.lock [hunk 3]
```diff
@@ -1,67 +1,138 @@
...
       fog (~> 1.0)
       rest-client (~> 1.0)
       unf (~> 0.1)
-    excon (0.52.0)
+    excon (0.55.0)
     fission (0.5.0)
       CFPropertyList (~> 2.2)
-    fog (1.25.0)
+    fog (1.38.0)
+      fog-aliyun (>= 0.1.0)
```
Comment (BenFradet) [inline 3-enrich/emr-etl-runner/Gemfile.lock]: From what I understand between those two versions of fog the project got splitted in multiple gems and requiring fog requires everything.

File: 3-enrich/emr-etl-runner/lib/snowplow-emr-etl-runner/cli.rb [hunk 4]
```diff
@@ -41,101 +50,142 @@ module Cli
...
+
+        commands = {
+          'run' => OptionParser.new do |opts|
+            opts.banner = 'Usage: run [options]'
+            opts.description = 'Run the Snowplow pipeline on EMR'
+            opts.on('-c', '--config CONFIG', 'configuration file') { |config| options[:config_file] = config }
+            opts.on('-n', '--enrichments ENRICHMENTS', 'enrichments directory') { |config| options[:enrichments_directory] = config }
+            opts.on('-r', '--resolver RESOLVER', 'Iglu resolver file') { |config| options[:resolver_file] = config }
+            opts.on('-d', '--debug', 'enable EMR Job Flow debugging') { |config| options[:debug] = true }
+            opts.on('-f', '--resume-from {enrich,shred,elasticsearch,archive_raw}', 'resume from the specified step') { |config| options[:resume_from] = config }
```
Comment (BenFradet) [inline 3-enrich/emr-etl-runner/lib/snowplow-emr-etl-runner/cli.rb]: good idea

File: 3-enrich/emr-etl-runner/Gemfile.lock (hunk 3 above)
Comment (chuwy) [inline 3-enrich/emr-etl-runner/Gemfile.lock]: Got it.

Comment (BenFradet) [inline 3-enrich/emr-etl-runner/Gemfile.lock]: And Elasticity [requires fog](https://github.com/rslifka/elasticity/blob/master/elasticity.gemspec#L16).

File: 3-enrich/emr-etl-runner/lib/snowplow-emr-etl-runner/generator.rb [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        else
+          raise ConfigError, "Config could not be validated against the schema #{avro_schema}"
+        end
+        nil
+      end
+
+      # Get the associated Avro schema
+      Contract String => String
+      def get_schema(version)
+        url = "https://raw.githubusercontent.com/snowplow/iglu-central/master/schemas/" +
```
Comment (alexanderdean) [inline 3-enrich/emr-etl-runner/lib/snowplow-emr-etl-runner/generator.rb]: Agree, we should use the resolver (that's what it's for)

Comment (BenFradet) [inline 3-enrich/emr-etl-runner/lib/snowplow-emr-etl-runner/generator.rb]: Yup I'll do that, I went this route not to have to supply a resolver file to generate an emr-config file but it doesn't really seem worth it.

File: 3-enrich/emr-etl-runner/lib/snowplow-emr-etl-runner/emr_config_generator.rb [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+        get_schema_name("ClusterConfig", "avro", version)
+      end
+
+      Contract ConfigHash, Bool, Maybe[String], String, ArrayOf[String] => Hash
+      def create_datum(config, debug=false, resume_from=nil, resolver='', enrichments=[])
+        legacy = (not (config[:aws][:emr][:ami_version] =~ /^[1-3].*/).nil?)
+        region = config[:aws][:emr][:region]
+
+        {
+          "name" => config[:enrich][:job_name],
```
Comment (BenFradet) [inline 3-enrich/emr-etl-runner/lib/snowplow-emr-etl-runner/emr_config_generator.rb]: The problem is that using symbolic keys, the json fails Avro validation.

Comment (chuwy) [inline 3-enrich/emr-etl-runner/lib/snowplow-emr-etl-runner/emr_config_generator.rb]: Does it mean `Avro::Schema.validate` requires keys to be strings?

Comment (BenFradet) [inline 3-enrich/emr-etl-runner/lib/snowplow-emr-etl-runner/emr_config_generator.rb]: It does seem to be the case, yes

Comment (chuwy) [inline 3-enrich/emr-etl-runner/lib/snowplow-emr-etl-runner/emr_config_generator.rb]: Welp. That's ok then.

**Comment by BenFradet** [comment]
I think I've taken care of everything you mentioned @chuwy :+1:

File: 3-enrich/emr-etl-runner/Gemfile.lock (hunk 3 above)
Comment (BenFradet) [inline 3-enrich/emr-etl-runner/Gemfile.lock]: Building on your remark, I think I'll open a PR against elasticity specifying that we only depend on `fog-aws` that'll avoid all those unnecessary deps.

Comment (BenFradet) [inline 3-enrich/emr-etl-runner/Gemfile.lock]: opened rslifka/elasticity/pull/140

File: 3-enrich/emr-etl-runner/Gemfile [hunk 7]
```diff
@@ -21,13 +21,14 @@ ruby "2.3.1"
 # RubyGems it requires here.
 gem "contracts", "~> 0.9", "<= 0.11"
 gem "elasticity", "~> 6.0.12"
-gem "sluice", "~> 0.4.0"
+gem "avro", "~> 1.8.1"
 gem "awrence", "~> 0.1.0"
 gem "snowplow-tracker", "~> 0.5.2"
 gem "iglu-ruby-client", ">= 0.1.0"
+gem "sluice", "~> 0.4.0"
```
Comment (alexanderdean) [inline 3-enrich/emr-etl-runner/Gemfile]: Why move sluice to the bottom?

File: 3-enrich/emr-etl-runner/lib/snowplow-emr-etl-runner/runner.rb [hunk 8]
```diff
@@ -52,48 +49,43 @@ def initialize(args, config, enrichments_array, resolver, targets_array)
...
+        resume = @args[:resume_from]
+        steps = {
+          :staging => resume.nil?,
+          :enrich => (resume.nil? or resume == 'enrich'),
+          :shred => (resume.nil? or [ 'enrich', 'shred' ].include?(resume)),
+          :es => (resume.nil? or [ 'enrich', 'shred', 'elasticsearch' ].include?(resume)),
+          :archive_raw => (resume.nil? or [ 'enrich', 'shred', 'elasticsearch', 'archive_raw' ].include?(resume)),
+          :rdb_load => (resume.nil? or [ 'enrich', 'shred', 'elasticsearch', 'archive_raw', 'rdb_load' ].include?(resume)),
+          :analyze => (resume.nil? or [ 'enrich', 'shred', 'elasticsearch', 'archive_raw', 'rdb_load', 'analyze' ].include?(resume))
+        }
```
Comment (BenFradet) [inline 3-enrich/emr-etl-runner/lib/snowplow-emr-etl-runner/runner.rb]: @chuwy flagging this as I'm not entirely sure of the order of the steps

File: 3-enrich/emr-etl-runner/Gemfile (hunk 7 above)
Comment (BenFradet) [inline 3-enrich/emr-etl-runner/Gemfile]: I had removed it and had to add it back

File: 3-enrich/emr-etl-runner/lib/snowplow-emr-etl-runner/cli.rb [hunk 9]
```diff
@@ -42,106 +52,146 @@ module Cli
...
         end
 
         # Run OptionParser's structural validation
         begin
-          optparse.parse!
+          global.order!
+          cmd_name = ARGV.shift
+          if cmd_name
+            cmd = commands[cmd_name]
+            if cmd
```
Comment (BenFradet) [inline 3-enrich/emr-etl-runner/lib/snowplow-emr-etl-runner/cli.rb]: nothing regarding subcommand unfortunately

**Comment by alexanderdean** [comment]
Hey, a quick bit of feedback on rc2:

* Calling `emr-etl-runner --config` (i.e. forgetting the `run` command)
* gives "invalid option" not "invalid command"
* which leads me to think there's a bug in the handling code
* which is a pretty hard-to-read mix of exception-bubble-up versus local `exit1()` helper function

**Comment by BenFradet** [comment]
@alexanderdean 

The handling isn't at fault here, you've effectively specified an invalid option, the only global option being `--version`. Moreover, this handling is done by `OptionParser` so I have no control over it.

**Comment by alexanderdean** [comment]
Got it, so `emr-etl-runner config` *would* have triggered invalid command, but `emr-etl-runner --config` is treated as an invalid global option.

**Comment by BenFradet** [comment]
exactly :+1:

**Comment by alexanderdean** [comment]
Please take a look at this one: https://snowplow.pagerduty.com/incidents/P9WTFX2

I don't know if this is a R90-era bug, or a R91 regression, but we shouldn't be trying and failing to download the RDB log if the job failed at a prior step - because there will never be a log. /cc @chuwy

**Comment by alexanderdean** [comment]
I have manually added `snowplow-hadoop-fs-rmr.sh` to snowplow-hosted-assets to unblock testing, but this isn't ideal, and I don't intend to manually copy it to all the other buckets.

Can we please:

* Version this file like everything else for invocation (so this filename would be deployed as `snowplow-hadoop-fs-rmr-0.1.0-rc1.sh` and invoked as such)
* Add it to Release Manager so that it is auto-deployed to all buckets

**Comment by alexanderdean** [comment]
Can we harmonize the step names a bit? There's a lot of inconsistency.

Currently:

```
- 1. Elasticity S3DistCp Step: Staging of s3n://some-logs-snplow/
- 2. Elasticity S3DistCp Step: Raw S3 -> HDFS
- 3. Elasticity Spark Step: Enrich Raw Events
- 4. Elasticity S3DistCp Step: Enriched HDFS -> S3
- 5. Elasticity S3DistCp Step: Enriched HDFS _SUCCESS -> S3
- 6. Elasticity Custom Jar Step: Recursively removing content from hdfs:///local/snowplow/raw-events/
- 7. Elasticity Spark Step: Shred Enriched Events
- 8. Elasticity S3DistCp Step: Shredded HDFS _SUCCESS -> S3
- 9. Elasticity S3DistCp Step: Shredded HDFS -> S3
- 10. Elasticity Custom Jar Step: Load Snowplow Redshift eu-west-1 Storage Target
- 11. Elasticity S3DistCp Step: Shredded S3 -> S3 Shredded Archive
- 12. Elasticity S3DistCp Step: Enriched S3 -> S3 Enriched Archive
- 13. Elasticity S3DistCp Step: Raw S3 Staging -> S3 Archive
```

Let's rename the first two steps:

```
- 1. Elasticity S3DistCp Step: Raw s3n://some-logs-snplow/ -> Raw Staging S3
- 2. Elasticity S3DistCp Step: Raw Staging S3 -> Raw HDFS
```

Simplify the name of the recursive delete:

```
- 6. Elasticity Custom Jar Step: Clear Raw HDFS
```

For consistency, put "S3" last in the final three copies, and make more consistent:

```
- 11. Elasticity S3DistCp Step: Shredded S3 -> Shredded Archive S3
- 12. Elasticity S3DistCp Step: Enriched S3 -> Enriched Archive S3
- 13. Elasticity S3DistCp Step: Raw Staging S3 -> Raw Archive S3
```

This should make everything much less artisanal.

File: 3-enrich/emr-etl-runner/lib/snowplow-emr-etl-runner/emr_config_generator.rb [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+          "tags" => get_tags(config[:monitoring][:tags]),
+          "bootstrapActionConfigs" => get_bootstrap_actions(
+            config[:aws][:emr][:bootstrap],
+            config[:collectors][:format],
+            legacy,
+            region,
+            config[:enrich][:versions][:spark_enrich]
+          ),
+          "configurations" => get_configurations(legacy),
+          "applications" => ["Hadoop"]
```
Comment (BenFradet) [inline 3-enrich/emr-etl-runner/lib/snowplow-emr-etl-runner/emr_config_generator.rb]: There wasn't Spark enrich/shred at the time. Anyway the generate command needs to support r89 and r90.

File: 3-enrich/emr-etl-runner/lib/snowplow-emr-etl-runner/emr_config_generator.rb [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+      def create_datum(config, debug=false, resume_from=nil, resolver='', enrichments=[])
+        legacy = (not (config[:aws][:emr][:ami_version] =~ /^[1-3].*/).nil?)
+        region = config[:aws][:emr][:region]
+
+        {
+          "name" => config[:aws][:emr][:jobflow][:job_name],
+          "logUri" => config[:aws][:s3][:buckets][:log],
+          "region" => region,
+          "credentials" => {
+            "accessKeyId" => config[:aws][:access_key_id],
```
Comment (BenFradet) [inline 3-enrich/emr-etl-runner/lib/snowplow-emr-etl-runner/emr_config_generator.rb]: this is handled by df-runner: https://github.com/snowplow/dataflow-runner/blob/master/src/utils.go#L25-L41

File: 3-enrich/emr-etl-runner/lib/snowplow-emr-etl-runner/emr_config_generator.rb (hunk 10 above)
Comment (BenFradet) [inline 3-enrich/emr-etl-runner/lib/snowplow-emr-etl-runner/emr_config_generator.rb]: logged #3359

**Comment by BenFradet** [comment]
@chuwy what do you want to do regarding rdb loader logs? Should we just retrieve them only if the emr job was successful?

**Comment by BenFradet** [comment]
@alexanderdean the problem seems to be that we want the logs especially if rdb loader failed and we have, at the moment, no way to know which step failed.

**Comment by alexanderdean** [comment]
But EmrEtlRunner retrieves the statuses of each jobflow step - doesn't that tell us whether the RDB Loader step failed or not?

**Comment by BenFradet** [comment]
pushed an rc4 taking care of hopefully everything and an rc1 of `s3://snowplow-hosted-assets/common/emr/snowplow-hadoop-fs-rmr-0.1.0-rc1.sh`

**Comment by alexanderdean** [comment]
Thanks @BenFradet !

**Pull request merged**